### PR TITLE
Ultimate TTT AI

### LINKED ISSUE #267: Ultimate TTT: Empty squares are sometimes impossible to navigate to
**Description**
If the bottom-right square in a subgame is empty, but the rest of the bottom row and right column are filled, the cursor cannot be moved from the top-left section to the bottom-right square.

**Expected behavior**
You should somehow be able to move down and to the right simultaneously. This seems to be possible already for the top left and right corners 

**Reproduction Steps**
_Steps to reproduce the behavior and how often it occurs._
1. Fill up the rightmost  column and the bottom row with markers, except for the bottom-right corner.
2. Try to select the corner

**Screenshots**
See this image for an example of the situation, outlined in white - the square with the white circle cannot be accessed:
cant-move

### PR DESCRIPTION
**Pull request by dylwhich**
### Description

Adds a single-player CPU opponent with 3 difficulties:

* Easy -- Plays a random move
* Medium -- Plays optimally within sub-games, but has no overall strategy
* Hard -- Analyzes the board and tries to play the best overall move, taking into account the best sub-game move, and weighing that against what the opponent's best move would be in the sub-game they would get to move in. Or, if they would get to pick any sub-game, their best move overall.

Also changes `tttHandleGameInput()` to use a slightly different method, which should prevent any squares from becoming inaccessible and I think maybe it handles a bit more intuitively than the previous one? I recommend playing with it to see. But it pretty much tries to find a close free space before it wraps all the way around by bumping the cursor away from the edge of the board, and it also prioritizes the last intended direction for the secondary cursor direction.

### Test Instructions

Play against the AIs, see how they feel. See if the cursor behaves reasonably in some funky situations.

### Ticket Links

Closes #267 

### Readiness Checklist
- [x] I have run `make format` to format the changes
- [x] I have compiled the firmware and the changes have no warnings
- [x] I have compiled the emulator and the changes have no warnings
- [x] I have run `make cppcheck` and checked that `cppcheck_result.txt` has no warnings for the changes
- [x] I have added doxygen comments to any code used by more than one Swadge mode. This includes `/*! \file` comments with Design Philosophy, Usage, and Example sections for new headers.
- [x] I have run `make docs` and checked that `doxy_warnings.txt` has no warnings for the new code

### CODE REVIEW

**Comment by AEFeinstein** [comment]
@dylwhich I stapled three things onto this PR (hope you don't mind):
1. More markers!
1. Randomize play order for single player
1. Randomize CPU marker to be an unlocked one which isn't the players

**Comment by dylwhich** [comment]
Nice, good idea! Hah, I think I might have hardcoded the Hard CPU to make the move that's most advantageous for Player 2, because it really feels like it's _avoiding_ winning against me when it goes first

File: main/modes/games/ultimateTTT/ultimateTTTgame.c [hunk 1]
```diff
@@ -200,33 +248,54 @@ void tttHandleGameInput(ultimateTTT_t* ttt, buttonEvt_t* evt)
...
             {
-                cursorFunc          = incCursorX;
-                cursorFuncSecondary = incCursorY;
+                cursorFunc                = incCursorX;
+                cursorFuncSecondary       = CHOOSE_CURSOR_Y();
+                ttt->game.cursorLastDir.x = 1;
                 break;
             }
+
+#undef CURSOR_DIR_X
```
Comment (AEFeinstein) [inline main/modes/games/ultimateTTT/ultimateTTTgame.c:294]: Do you really need these `#undef`s?

File: main/modes/games/ultimateTTT/ultimateTTTgame.c [hunk 2]
```diff
@@ -593,42 +683,47 @@ static void tttPlaceMarker(ultimateTTT_t* ttt, const vec_t* subgame, const vec_t
...
+        if (!ttt->game.singlePlayer)
         {
-            // Increment wins
-            ttt->wins++;
-            writeNvs32(tttWinKey, ttt->wins);
-            ttt->lastResult = TTR_WIN;
-
-            // Check for unlocked markers
-            for (int16_t mIdx = 0; mIdx < NUM_UNLOCKABLE_MARKERS; mIdx++)
+            // Don't give points for beating the computer!
```
Comment (AEFeinstein) [inline main/modes/games/ultimateTTT/ultimateTTTgame.c:688]: I was wondering about that. I'd probably allow unlocks against CPU on the principle that stuff should be unlockable without needing multiple Swadges.

File: main/modes/games/ultimateTTT/ultimateTTTgame.c [hunk 3]
```diff
@@ -593,42 +683,47 @@ static void tttPlaceMarker(ultimateTTT_t* ttt, const vec_t* subgame, const vec_t
...
+            // At least, not until it has a hard mode
```
Comment (AEFeinstein) [inline main/modes/games/ultimateTTT/ultimateTTTgame.c:689]: Hm, this exists now...

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c [hunk 4]
```diff
@@ -0,0 +1,1192 @@
...
+
+static bool selectSubgame_medium(ultimateTTT_t* ttt, int* x, int* y)
+{
+    // Still just random
+    return selectSubgame_easy(ttt, x, y);
+}
+
+static bool selectSubgame_hard(ultimateTTT_t* ttt, int* x, int* y)
+{
+    // Eventually this will become the medium difficulty once I get the actual difficult AI working
```
Comment (AEFeinstein) [inline main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:311]: Is this still pending?

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c [hunk 5]
```diff
@@ -0,0 +1,1192 @@
...
+    {
+        rowCount_t rowCount  = checkDiag(subgame, n, player);
+        uint16_t movesNeeded = 3 - TO_PLAYERS(rowCount);
+        if (0 == TO_OPPONENTS(rowCount) && movesNeeded < minMovesToWin)
+        {
+            minMovesToWin = movesNeeded;
+        }
+    }
+
+    return minMovesToWin;
```
Comment (AEFeinstein) [inline main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:843]: Code after this is dead. Intentional?

**Review by AEFeinstein — APPROVED** [review]

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c (hunk 5 above)
Comment (dylwhich) [reply main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:843]: Ah, no I should delete that. Surprised it doesn't generate a warning

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c (hunk 4 above)
Comment (dylwhich) [reply main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:311]: Hmm, well I still do want to do this but I'm not sure if I can commit to it until I get some time to actually work on it... I think the current "Hard" AI is at least a worthy opponent in that it doesn't make super obvious errors and it plays well enough to make it interesting. But it doesn't actually feel that hard to me, so I still do want to try and write a more advanced AI that actually thinks a few moves ahead, and then either have parameters to tune its difficulty for the other levels, or shift everything down a level and drop the pure-random difficulty.

File: main/modes/games/ultimateTTT/ultimateTTTgame.c (hunk 2 above)
Comment (dylwhich) [reply main/modes/games/ultimateTTT/ultimateTTTgame.c:688]: That's fair. Should it be for all difficulties, or just hard maybe?

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c (hunk 4 above)
Comment (AEFeinstein) [reply main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:311]: Ah, gotcha. I'll consider that a stretch goal, but not a top priority.

File: main/modes/games/ultimateTTT/ultimateTTTgame.c (hunk 2 above)
Comment (AEFeinstein) [reply main/modes/games/ultimateTTT/ultimateTTTgame.c:688]: I'd tilt towards all difficulties. That way we don't punish young'ns who can play on easy. Mensa members can still play hard, and I don't think they'd mind (or notice) that easy also unlocks.

**Comment by AEFeinstein** [comment]
This all looks good to me (and feels good too). You can merge after addressing feedback (if you wanna)

File: main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c (hunk 5 above)
Comment (AEFeinstein) [reply main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c:843]: Yeah, not even `cppcheck` caught it. Odd.

File: main/modes/games/ultimateTTT/ultimateTTTgame.c (hunk 1 above)
Comment (dylwhich) [reply main/modes/games/ultimateTTT/ultimateTTTgame.c]: Nah not really! I'll `#unundef` them

File: main/modes/games/ultimateTTT/ultimateTTTgame.c (hunk 2 above)
Comment (dylwhich) [reply main/modes/games/ultimateTTT/ultimateTTTgame.c]: Done!

**Pull request merged**